File: solution/34/solution.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if len(nums) < 1:return [-1, -1]
        left = self.searchLeft(nums, target)
        return [left, self.searchRight(left, nums, target)]

    def searchLeft(self, nums, target):
        begin = 0
        end = len(nums) - 1
        while begin + 1 < end:
            mid = (begin + end) / 2
            if nums[mid] >= target:
                end = mid
            else:
                begin = mid
        if nums[begin] == target:return begin
        elif nums[end] == target: return end
        return -1

    def searchRight(self, left, nums, target):
        if left == -1:
            return -1
        if left == len(nums) - 1:
            return left
        begin = left
        end = len(nums) - 1
        while begin+1 < end:
            mid = (begin + end) / 2
            if nums[mid] > target:
                end = mid
            else:
                begin = mid
        if nums[end] == target:return end
        elif nums[begin] == target: return begin
        return -1

File: solution/34/solution.cpp
#include <stdio.h>
#include <vector>

class Solution {
public:
    std::vector<int> searchRange(std::vector<int>& nums, int target) {
        int left = searchLeft(nums, target);
        int right = searchRight(nums, target);
        std::vector<int> result;
        result.push_back(left);
        result.push_back(right);
        return result;
    }

    int searchLeft(std::vector<int> &nums, int target){
        int left = 0;
        int right = nums.size() - 1;
        while(left <= right){
            int mid = (left + right)/2;
            if(nums[mid] == target){
                if(mid==0||nums[mid-1]!=target)return mid;
                right = mid - 1;
            }
            else if(nums[mid] > target) right = mid -1;
            else left = mid + 1;
        }
        return -1;
    }

    int searchRight(std::vector<int> &nums, int target){
        int left = 0;
        int right = nums.size() - 1;
        while(left <= right){
            int mid = (left + right)/2;
            if(nums[mid] == target){
                if(mid==nums.size()-1||nums[mid+1]!=target)return mid;
                left = mid + 1;
            }
            else if(nums[mid] > target)right = mid -1;
            else left = mid + 1;
        }
        return -1;
    }
};

int main(){
    Solution s;
}